Modernize the application & Performance tweaks

khaledosman opened the pull request:
The repo seems to be a bit out of date and isn't using best practices when it comes to performance for React applications, so this article might be unrealistic for 2019 https://medium.com/free-code-camp/a-realworld-comparison-of-front-end-frameworks-with-benchmarks-2019-update-4be0d3c78075. This PR aims to change that 

- Used the newest versions of React and other dependencies
- Migrate away from deprecated library react-router-redux to connected-react-router
- Use PureComponents and React.memo for functional components for better performance
- Migrate away from deprecated lifeCycleHooks like componentWIllUpdate, componentWillReceiveProps & componentWillMount
- lazyload & codesplit non-default routes with React.lazy & React.Suspense

khaledosman committed:
update dependencies, migrate from willMount to didMount, migrate from react-router-redux to connected-react-router, use PureComponents/memo

khaledosman committed:
update react-redux to v7

khaledosman committed:
lazy-loading & codeplitting

khaledosman committed:
bump dependencies

khaledosman committed:
update to create-react-app/react-scripts v3

khaledosman committed:
package-lock

khaledosman committed:
get rid of unnecessary componentWillReceiveProps in Settings component

khaledosman committed:
migrate away from deprecated componentWillReceiveProps

khaledosman committed:
update deprecated import

sangdth commented:
I have the same feeling and looking forward to this PR.

khaledosman committed:
bump dependencies

khaledosman committed:
bump dependencies

khaledosman committed:
upgrade to react-router to v5.0.1

khaledosman commented:
Any updates?

khaledosman committed:
update react-redux to v7.1.0

klamping commented:
FYI, using your fork, logging out gives me this error:
![image](https://user-images.githubusercontent.com/706039/59813806-8bf21480-92d8-11e9-90e8-772e9ad9854f.png)

khaledosman commented:
@klamping steps to reproduce? I just run it locally and it works fine.

khaledosman committed:
bump dependencies

khaledosman committed:
fix error on logout

khaledosman commented:
@klamping I was able to reproduce the issue, it should be fixed now.

khaledosman committed:
update marked to v0.7.0

khaledosman committed:
fix change between uncontrolled to controlled component warnings

khaledosman commented:
Is anyone maintaining this app anymore? I would be happy to maintain it myself

khaledosman committed:
update to react v16.9.0

khaledosman committed:
fix warnings

khaledosman committed:
use fontdisplay:swap for google fonts

helderscrolls commented:
@khaledosman One small thing I found, currentUser.image is broken if you define no image, isn't there a way already coded on the API to recover the profile.image smiley face and use it instead of displaying the alt message ?

khaledosman commented:
@helderscrolls Thanks, I'm not sure where that broke as I didn't change any API related logic, I'll look into it later when I have the time.

helderscrolls commented:
@khaledosman I think it’s a issue that comes from the original Build, the only way I’ve found so far is to hard code the URL on the front-end, on the Header and the Comment preview that way even if currentUser.image is undefined it displays the smiley face

khaledosman committed:
bump dependencies

khaledosman committed:
fix: set default user image

khaledosman commented:
@helderscrolls This should be fixed now, I've added a fallback default image everywhere that is using `user.image`

dabonnestor commented:
There is a bug when trying to submit an article. To reproduce the bug try to register as a new user then submit an article without inputing a data on the fields.

![Screen Shot 2019-09-18 at 4 05 14 PM](https://user-images.githubusercontent.com/22286294/65129584-2b38bc00-da2e-11e9-9746-00e4d3457e16.png)

khaledosman commented:
@dabonnestor looks like validation is not implemented for this form, can you check if this works on master branch?

dabonnestor commented:
It has validation on the master branch.

khaledosman committed:
fix: ARTICLE_SUBMITTED action causes error when article is null

khaledosman commented:
this has been fixed.

dabonnestor commented:
You should be the active maintainer for this repo 👍

shanebdavis commented:
I also forked this project - to convert it all to Hooks. Then I noticed your excellent work here, @khaledosman. Lesson learned! I should check the existing forks before starting my own! I can't believe this has been sitting here for 6 months! 

My hooks conversion went smoothly. I also attempted to do the updates you did (redux, deprecated react-router-redux, etc), but I ran into complications, so I kept my fork focused.

Just browsing your source, I don't see that you converted things to hooks, though you modernized everything else. Would you be interested in my attempting to merge my work with yours?

khaledosman commented:
@shanebdavis absolutely go for it, you could make a branch out of mine.

I don't see the point if the contributors will ignore it anyway though.

shanebdavis commented:
> I don't see the point if the contributors will ignore it anyway though.

I see that the main real-world repo is accepting pull-requests. I'll try to bug them over there.

khaledosman committed:
chore: bump dependencies

khaledosman committed:
fix: bump dependencies

khaledosman committed:
Merge branch 'master' of https://github.com/khaledosman/react-redux-realworld-example-app

khaledosman committed:
chore: bump dependencies

khaledosman committed:
chore: bump dependencies

khaledosman committed:
chore: bump dependencies

khaledosman committed:
bump dependencies

ezorfa commented:
@khaledosman There is this error when i click signin / settings or signup 👍 
"Uncaught Could not find router reducer in state tree, it must be mounted under "router""

khaledosman commented:
@ezorfa I couldn't reproduce this error, can you delete node_modules/ package-lock.json or yarn.lock and try with a clean npm install?

it seems the backend API doesn't support CORS anymore so I'm getting this error instead
```
localhost/:1 Access to XMLHttpRequest at 'https://conduit.productionready.io/api/articles?limit=10&offset=0' from origin 'http://localhost:4100' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
```
I added some checks to make the app render when the api returns an error, it seems like error states are not implemented though.

khaledosman committed:
dont break the app if the api request fails

khaledosman committed:
downgrade history to v4.10

ezorfa commented:
@khaledosman  which backed do you use?

ezorfa commented:
@khaledosman  I get this error now:

<img width="870" alt="Screenshot 2020-08-28 at 10 31 29" src="https://user-images.githubusercontent.com/50787308/91539458-a6327400-e919-11ea-8429-fbda2cea078e.png">

Console:

<img width="539" alt="Screenshot 2020-08-28 at 10 31 52" src="https://user-images.githubusercontent.com/50787308/91539488-b21e3600-e919-11ea-8b36-9bb7be0cb307.png">


I am using django backend.

khaledosman commented:
@ezorfa I'm not sure, its whatever backend the app was using before, I think they all use the same backend because I got the same issue in the stencil-realworld-app aswell which uses the same endpoint https://github.com/khaledosman/stencil-realworld-app

The issue you saw is the one with the error state I just mentioned, please do a git pull to get my latest commit

sashee committed:
add backend URL environment parameter

sashee committed:
Add documentation for the REACT_APP_BACKEND_URL

khaledosman committed:
Merge pull request #1 from sashee/master

Make the API_ROOT configurable with an environment variable

khaledosman committed:
update to react17 and cra v4

sashee committed:
Fix missing image in Header

khaledosman committed:
Merge pull request #2 from sashee/patch-1

Fix missing image in Header

khaledosman committed:
bump dependencies

leosuncin committed:
replace marked with snarkdown

Install xss to sanitize untrusted HTML

leosuncin committed:
Configure the store with Redux Toolkit

leosuncin committed:
remove superagent dependency

Use native fetch to make API request.

leosuncin committed:
convert article reducer to slice

Use async thunk actions and action creators.

leosuncin committed:
convert auth reducer to slice

Use async thunk actions and action creators.

leosuncin committed:
Fix ListError component

Import the correct component

leosuncin committed:
convert settings reducer to slice

Use async thunk actions and action creators.
Use createAction for logout action.

leosuncin committed:
convert profile reducer to slice

Use async thunk actions and action creators.

TODO: handle getByAuthor async thunk action

leosuncin committed:
install and setup Cypress

Install @testing-library/cypress and @cypress/code-coverage

leosuncin committed:
add e2e test for home page

Check when is unathenticated.
Check the structure of the page.

leosuncin committed:
add e2e test for app navigation

Check when is unathenticated.
Navigate trought all the app.

leosuncin committed:
add e2e test for register page

Validate the form.
Install faker to generate test data.

leosuncin committed:
add e2e test for login page

Validate the form.
Add test user in Cypress' env.

leosuncin committed:
add e2e test for editor page

Allow to Cypress to access the Redux Store.
Validate the form.
Create a new article.
Add custom command to dispatch actions to the store.
Add custom command to login.

leosuncin committed:
add more e2e test for editor page

Edit an article.
Add custom command to create a new article using the API.

leosuncin committed:
add e2e test for settings page

leosuncin committed:
add e2e test for article page

Create a new comment
Delete a comment
Remove an article

TODO: fix add new comment at the beginning of the list
TODO: fix show validation errors

leosuncin committed:
generate a markdown with faker

Navigate to edit the article.
Do not download a random markdown.

leosuncin committed:
add e2e test for profile page

Follow/unfollow a user.
Add a task to generate an user and an article.

TODO(backend): fix get articles by author for a different user than authenticated user.

leosuncin committed:
add more e2e test for app navigation

Check when is authenticated.

leosuncin committed:
add more e2e test for home page

Check when is authenticated.
Mark an article as favorite/unfavorite.

leosuncin committed:
fix generated markdown

leosuncin committed:
stop using markdown because causes issues with compare strings

leosuncin committed:
ignore .nyc_output directory

leosuncin committed:
Merge branch 'e2e-tests' into redux-toolkit

# Conflicts:
#	package-lock.json
#	package.json

leosuncin committed:
fix generate valid values

leosuncin committed:
Merge branch 'e2e-tests' into redux-toolkit

leosuncin committed:
fix login custom command

leosuncin committed:
fix comment actions

Add new comment at the beginning.
Show validation errors.
Remove a comment.

leosuncin committed:
add more e2e test for article page

Add new comment at the beginning.
Show validation errors.

leosuncin committed:
remove editor reducer

Handle the state of Editor component locally.
Add more async thunk actions for article.

TODO: there are some bugs when navigate to "New Post" from an article.

leosuncin committed:
change the type of the action dispatched when create an article

leosuncin committed:
document API client methods

Fix return an error for an empty response.
Serialize query object to URL parameters.

leosuncin committed:
document API client methods

Fix return an error for an empty response.
Serialize query object to URL parameters.

leosuncin committed:
convert articleList reducer to slice

Remove home reducer.
Fix show all articles by author in profile page.
Change get all articles method of agent.

leosuncin committed:
convert articleList reducer to slice

Remove home reducer.
Fix show all articles by author in profile page.
Change get all articles method of agent.

leosuncin committed:
convert common reducer to slice

leosuncin committed:
convert common reducer to slice

leosuncin committed:
Remove promiseMiddleware

Refactor localStorageMiddleware

leosuncin committed:
Remove promiseMiddleware

Refactor localStorageMiddleware

khaledosman committed:
Update README.md

khaledosman committed:
update links

khaledosman committed:
Update README.md

khaledosman committed:
Merge remote-tracking branch 'khaled/master' into redux-toolkit

khaledosman committed:
merge & update app component to function component

khaledosman committed:
convert class components to function components

leosuncin committed:
fix fetch favorite articles in profile

leosuncin committed:
fix fetch favorite articles in profile

leosuncin committed:
regenerate NPM lock-file with npm v6

leosuncin committed:
regenerate NPM lock-file with npm v6

leosuncin committed:
fix load article in editor

leosuncin committed:
fix load article in editor

leosuncin committed:
Merge branch 'khaledosman-redux-toolkit' into redux-toolkit

# Conflicts:
#	package-lock.json
#	src/agent.js
#	src/components/App.js
#	src/components/ArticlePreview.js
#	src/components/Editor.js
#	src/components/Home/index.js
#	src/components/Profile.js
#	src/components/ProfileFavorites.js

leosuncin committed:
fix syntax error in articleList

leosuncin committed:
refactor login component to use hooks instead of connect

Add some JSDoc tags to set typing.

leosuncin committed:
refactor list error component

Document the component

leosuncin committed:
refactor register component to use hooks instead of connect

Document the component.

leosuncin committed:
refactor settings components to use hooks instead of connect

Document the components.

leosuncin committed:
fix profile screen component

Use hooks instead of connect.
Refactor profile component.
Profile and Favorites page use the same component.

leosuncin committed:
refactor editor component to use hooks instead of connect

Document the component.

FIXES load article only when edit and clear the form after leave the screen

leosuncin committed:
refactor article component to use hooks instead of connect

Document all components.
Prefer use hooks over prop drilling.

BREAKING CHANGE lazy load the CommentContainer, handle the case when article is loading or not exists

leosuncin committed:
refactor to avoid prop drilling in ArticleList

Use hooks over props.
Document all components.

leosuncin committed:
refactor home component to use hooks instead of connect

Document all components.
Prefer use hooks over prop drilling.
Do not override tags when fetch articles by tag.

BREAKING CHANGE tags are fetched from tag component not home component

leosuncin committed:
refactor app component to use hooks instead of connect

Document all components.

leosuncin committed:
chore: setup git hooks

leosuncin committed:
refactor(tags): add tags slice

leosuncin committed:
test(tags): add unit tests

leosuncin committed:
refactor(tags): sidebar component

leosuncin committed:
refactor: add function to create the store

Help to recreate the store inside the tests.

leosuncin committed:
chore: setup tests

leosuncin committed:
test(tags): sidebar component

Add custom render function, wrap the providers.

leosuncin committed:
refactor(tags): tag list component

leosuncin committed:
refactor(auth): auth slice

Move all authentication state to it.

BREAKING CHANGE removed loginPageUnloaded and registerPageUnloaded

leosuncin committed:
test(auth): auth slice

Unit test the reducer and selectors

leosuncin committed:
refactor(auth): combine Login and Register into AuthScreen

Both Login and Register share almost all the UI, so, to reduce duplicate code.

BREAKING CHANGE: remove Login and Register components and use only one screen

leosuncin committed:
fix: call next action after the side effects

leosuncin committed:
test(auth): login and register forms

leosuncin committed:
refactor(settings): remove settings reducer

Now these features are under auth slice.

BREAKING CHANGE: remove settingsPageUnloaded

leosuncin committed:
refactor(auth): update settings component

Move to `auth` folder.

leosuncin committed:
test(auth): add tests for settings screen

leosuncin committed:
refactor(comments): add comments slice

Add a slice to manage the comments.

leosuncin committed:
test(comments): add tests for comments slice

leosuncin committed:
refactor(comments): arrange components

leosuncin committed:
test(comments): add tests for components

khaledosman committed:
Merge branch 'gothinkster:master' into master

khaledosman committed:
fix: replace url with api.realworld.io

Commit:
migrate to react v18, getrid of history & connected-react-router, upgrade react-router v6

Commit:
Merge remote-tracking branch 'leo/redux-toolkit' into upgrade-codebase

khaledosman committed:
Merge pull request #11 from khaledosman/upgrade-codebase

Upgrade codebase

Commit:
fix: fix register path matching

Commit:
fixes

Commit:
fixes

Commit:
pure components

Commit:
navigate after login/register

khaledosman committed:
update dependencies

noalbalint committed:
fix redirect handling on failed login

khaledosman committed:
Merge pull request #17 from noalbalint/fix-login-failure-flow

Fix redirect handling on failed login
